tapchannel/aux_leaf_creator.go

```diff
@@ -0,0 +1,282 @@
+package tapchannel
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/taproot-assets/address"
+	"github.com/lightninglabs/taproot-assets/fn"
+	cmsg "github.com/lightninglabs/taproot-assets/tapchannelmsg"
+	"github.com/lightningnetwork/lnd/channeldb"
+	lfn "github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+const (
+	// DefaultTimeout is the default timeout we use for RPC and database
+	// operations.
+	DefaultTimeout = 30 * time.Second
+)
+
+// LeafCreatorConfig defines the configuration for the auxiliary leaf creator.
+type LeafCreatorConfig struct {
+	ChainParams *address.ChainParams
+}
+
+// AuxLeafCreator is a Taproot Asset auxiliary leaf creator that can be used to
+// create auxiliary leaves for Taproot Asset channels.
+type AuxLeafCreator struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	cfg *LeafCreatorConfig
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewAuxLeafCreator creates a new Taproot Asset auxiliary leaf creator based on
+// the passed config.
+func NewAuxLeafCreator(cfg *LeafCreatorConfig) *AuxLeafCreator {
+	return &AuxLeafCreator{
+		cfg: cfg,
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}
+}
+
+// Start attempts to start a new aux leaf creator.
+func (c *AuxLeafCreator) Start() error {
```

Comment:
after a glimpse I don't get why does this need to be live?

Reply:
Hmm, you're right. Currently we don't require any state. But I think that might come with some later changes, so I'd like to keep it in (also to keep all the components looking and behaving similarly).